### PR TITLE
feat: adds authenticator support in factor select and challenge flow

### PR DESCRIPTION
**Pull request by haishengwu-okta**
## Description:

- 2nd commit about upgrade auth-js will be drop after backport of https://github.com/okta/okta-signin-widget/pull/1166

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:

- https://okta.box.com/s/kuaqe3gpozeyqxmimb5iluxa11wvsxs6

### Issue:

- [OKTA-288799](https://oktainc.atlassian.net/browse/OKTA-288799)

### CODE REVIEW

File: src/v2/models/AppState.js [hunk 1]
```diff
@@ -29,15 +29,19 @@ export default Model.extend({
 
   derived: {
     factorProfile: {
-      deps: ['factor'],
-      fn (factor = {}) {
-        return factor.profile || {};
+      // While we're moving toward `authenticator` platform, but still
+      // need to support `factor` for certain period.
+      // Could remove `factor` after it's deprecated completely.
+      deps: ['factor', 'authenticator'],
```
Comment (magizh-okta) [inline src/v2/models/AppState.js:35]: seems like views would still be referring to factorProfile which is a good idea for now but eventually I would call it authenticator instance.

Comment (haishengwu-okta) [reply src/v2/models/AppState.js:35]: I'm not sure if I understand your statement.
`factor` and `authenticator` won't exists at same time. how `view refer to factorProfile` if it doesn't exists. When API responds `authenticator`, you'll get `authenticator.profile`.

**Comment by haishengwu-okta** [comment]
> I am guessing we might need a overall flag / feature in appState that guides the views if we are in authenticator Context or factorProfile context.

Cannot tell for sure as I didn't see API contract for other use cases. but that won't be ideal if we have to do so. My hope is whether use `factor` or `authenticator` object shall be transparent to people building View for enrollment, verification etc.

File: test/testcafe/framework/page-objects/SelectFactorPageObject.js [hunk 2]
```diff
@@ -1,32 +1,73 @@
...
+  getFactorsCount() {
+    return this.form.getElement('.enroll-factor-list .enroll-factor-row').count;
+  }
+
+  getFactorLabelByIndex(index) {
+    return this.form.getElement('.enroll-factor-list .enroll-factor-row .enroll-factor-label').nth(index).textContent;
+  }
+
+  getFactorIconClassByIndex(index) {
+    return this.form.getElement('.enroll-factor-list .enroll-factor-icon-container .factor-icon').nth(index).getAttribute('class');
```
Comment (nikhilvenkatraman-okta) [inline test/testcafe/framework/page-objects/SelectFactorPageObject.js:19]: Should we declare these selectors as constants above as well ?

File: yarn.lock [hunk 3]
```diff
@@ -367,10 +367,10 @@
   version "1.1.3"
   resolved "https://registry.yarnpkg.com/@nodelib/fs.stat/-/fs.stat-1.1.3.tgz#2b5a3ab3f918cca48a8c754c08168e3f03eba61b"
 
-"@okta/okta-auth-js@^3.1.0":
-  version "3.1.0"
-  resolved "https://registry.yarnpkg.com/@okta/okta-auth-js/-/okta-auth-js-3.1.0.tgz#1d66868bfdc34e2969acf42f32fa238ac5ac4d78"
-  integrity sha512-6o0SSD6iH3T3vhUhAJByoPbpbLAfZl7uFz6S6If0gP872O39mrrj4spmEMtX8ygQheBnt23KFa7YZpVwAjB/9g==
+"@okta/okta-auth-js@3.1.1":
```
Comment (nikhilvenkatraman-okta) [inline yarn.lock:370]: why is this change included ?

**Comment by nikhilvenkatraman-okta** [comment]
> I am guessing we might need a overall flag / feature in appState that guides the views if we are in authenticator Context or factorProfile context.

It would be ideal if we can just use the `factor/factors` `authenticator/authenticators` object in the API and let the transformers do the job of making sure all the data is available for the views to render correctly.

File: yarn.lock (hunk 3 above)
Comment (haishengwu-okta) [reply yarn.lock:370]: seeing errors caused by https://github.com/okta/okta-signin-widget/pull/1166 during testcafe test. I'm not sure why it didn't happy in travis.
this commit will be removed after 4.0.2 backported.

File: playground/mocks/idp/idx/data/select-factor-authenticate-authenticators.json [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                                            "name": "methodType",
+                                            "required": false,
+                                            "value": "password",
+                                            "mutable": false
+                                        }
+                                    ]
+                                }
+                            },
+                            {
+                                "label": "Security Key or Biometric Authenticator (FIDO2)",
```
Comment (nikhilvenkatraman-okta) [inline playground/mocks/idp/idx/data/select-factor-authenticate-authenticators.json:43]: I did not see these options show up in your demo video. Are we handling the selection of any authenticator with this PR or just the password authenticator ?

Comment (haishengwu-okta) [reply playground/mocks/idp/idx/data/select-factor-authenticate-authenticators.json]: try to play with a few examples that we know the api structure.

File: webpack.playground.config.js [hunk 5]
```diff
@@ -94,6 +94,7 @@ module.exports = {
       const mockOptions = {
         multiRequest: false,
         proxy: false,
+        quiet: false,
```
Comment (nikhilvenkatraman-okta) [inline webpack.playground.config.js:97]: what does this option do ?

Comment (haishengwu-okta) [reply webpack.playground.config.js:97]: display more information to the console where mock server started.
```
okta-signin-widget git:(bacon-topic-add-authenticator-support) ✗ y start:dev-server --no-open
yarn run v1.21.1
$ webpack-dev-server --config webpack.playground.config.js --no-open
Registering POST service at /api/v1/authn/factors
Registering POST service at /api/v1/authn/factors/:factorid/lifecycle/activate
Registering POST service at /api/v1/authn/factors/:factorid/lifecycle/resend
Registering POST service at /api/v1/authn/factors/:factorid/verify
Registering POST service at /api/v1/authn/factors/:factorid/verify/resend
Registering POST service at /api/v1/authn
Registering POST service at /api/v1/authn/skip
Registering POST service at /idp/idx
Registering POST service at /idp/idx/authenticators/okta-verify/launch
Registering POST service at /idp/idx/authenticators/poll
Registering POST service at /idp/idx/authenticators/poll/cancel
Registering POST service at /idp/idx/authenticators/sso_extension/transactions/:transactionId/verify/cancel
Registering POST service at /idp/idx/cancel
Registering POST service at /idp/idx/challenge
Registering POST service at /idp/idx/challenge/answer
Registering POST service at /idp/idx/challenge/poll
Registering POST service at /idp/idx/challenge/resend
Registering POST service at /idp/idx/credential/enroll
Registering POST service at /idp/idx/enroll
Registering POST service at /idp/idx/identify
Registering POST service at /idp/idx/introspect
Registering POST service at /idp/idx/login/token/redirect
Registering POST service at /idp/idx/recover
Registering GET service at /idp/idx/authenticators/sso_extension/transactions/:transactionId/verify
Registering POST service at /idp/idx/authenticators/sso_extension/transactions/:transactionId/verify
Registering GET service at /login/getimage
Registering GET service at /login/sessionCookieRedirect
Project is running at http://localhost:3000/
webpack output is served from /
Content not from webpack is served from /Users/haishengwu/okta-oss/okta-signin-widget/playground, /Users/haishengwu/okta-oss/okta-signin-widget/target, /Users/haishengwu/okta-oss/okta-signin-widget/target/js, /Users/haishengwu/okta-oss/okta-signin-widget/target/css
Hash: 2e0e87fd4708f90dbd12
Version: webpack 3.12.0
Time: 1214ms
                   Asset    Size  Chunks                    Chunk Names
    playground.bundle.js  332 kB       0  [emitted]  [big]  playground.js
playground.bundle.js.map  395 kB       0  [emitted]         playground.js
   [2] multi (webpack)-dev-server/client?http://localhost:3000 ./playground/main.js 40 bytes {0} [built]
   [3] (webpack)-dev-server/client?http://localhost:3000 7.93 kB {0} [built]
   [4] ./node_modules/url/url.js 23.3 kB {0} [built]
   [5] ./node_modules/node-libs-browser/node_modules/punycode/punycode.js 14.7 kB {0} [built]
   [7] ./node_modules/url/util.js 314 bytes {0} [built]
   [8] ./node_modules/querystring-es3/index.js 127 bytes {0} [built]
  [11] ./node_modules/strip-ansi/index.js 161 bytes {0} [built]
  [12] ./node_modules/strip-ansi/node_modules/ansi-regex/index.js 135 bytes {0} [built]
  [13] ./node_modules/loglevel/lib/loglevel.js 7.86 kB {0} [built]
  [14] (webpack)-dev-server/client/socket.js 1.08 kB {0} [built]
  [16] (webpack)-dev-server/client/overlay.js 3.67 kB {0} [built]
  [21] (webpack)/hot nonrecursive ^\.\/log$ 170 bytes {0} [built]
  [23] (webpack)/hot/emitter.js 77 bytes {0} [built]
  [25] ./playground/main.js 1.5 kB {0} [built]
  [26] ./.widgetrc 262 bytes {0} [built]
    + 12 hidden modules
webpack: Compiled successfully.
Resolving response for POST /idp/idx/introspect
Resolving response for POST /idp/idx/challenge/poll
Resolving response for POST /idp/idx/challenge/poll
Resolving response for POST /idp/idx/challenge/poll
```

File: test/unit/spec/v2/ion/uiSchemaTransformer_spec.js [hunk 6]
```diff
@@ -1,13 +1,13 @@
 import { _ } from 'okta';
 import responseTransformer from 'v2/ion/responseTransformer';
 import uiSchemaTransformer from 'v2/ion/uiSchemaTransformer';
-import XHRFactorRequiredEmail from '../../../helpers/xhr/v2/FACTOR_REQUIRED_EMAIL';
-import XHREnrollProfile from '../../../helpers/xhr/v2/ENROLL_PROFILE';
-import XHRFactorEnrollOptions from '../../../helpers/xhr/v2/FACTOR_ENROLL_OPTIONS';
+import XHRFactorRequiredEmail  from '../../../../../playground/mocks/idp/idx/data/factor-verification-email.json';
```
Comment (magizh-okta) [inline test/unit/spec/v2/ion/uiSchemaTransformer_spec.js:4]: Add a test cases to cover authenticator cases ?

File: src/v2/ion/uiSchemaTransformer.js [hunk 7]
```diff
@@ -30,6 +47,55 @@ const createFactorTypeOptions = (options, factors) => {
...
+ * Example of the option like
+ * @param {AuthenticatorOption[]} options
+ * @param {AuthenticatorInstance[]} authenticators
+ */
+const createAuthenticatorOptions = (options = [], authenticators = []) => {
+  return options.map(option => {
+    const value = option.value && option.value.form && option.value.form.value || [];
+    const valueObject = value
+      // TODO:
+      // base on current API design, there maybe field that has
```
Comment (magizh-okta) [inline src/v2/ion/uiSchemaTransformer.js:80]: in which case does it come up?

Comment (haishengwu-okta) [reply src/v2/ion/uiSchemaTransformer.js:80]: Phone's method type as current proposal.
and in theory that's one possible combination.

File: webpack.playground.config.js (hunk 5 above)
Comment (nikhilvenkatraman-okta) [reply webpack.playground.config.js:97]: nice !

File: src/v2/util/FactorUtil.js [hunk 8]
```diff
@@ -32,7 +32,8 @@ const factorData = {
 };
 
 const getFactorData = function (factorName) {
-  return factorData[factorName];
+  const key = _.isString(factorName) ? factorName.toLowerCase() : '';
```
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/v2/util/FactorUtil.js:35]: We did. the label will be override by what comes back from API.

File: src/v2/view-builder/internals/FormInputFactory.js [hunk 9]
```diff
@@ -19,10 +19,26 @@ const createFactorTypeView = (opt) => {
...
   };
 };
+
+const createAuthenticatorSelectView = (opt) => {
+  var optionItems = (opt.options || [])
+    .map(opt => {
+      return Object.assign({}, FactorUtil.getFactorData(opt.authenticatorType), opt);
+    });
+  return {
+    View: FactorOptions,
```
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/v2/view-builder/internals/FormInputFactory.js:29]: yeah, there are still many things need to refactoring, like file name, class name.

File: test/unit/spec/v2/ion/uiSchemaTransformer_spec.js (hunk 6 above)
Comment (haishengwu-okta) [reply test/unit/spec/v2/ion/uiSchemaTransformer_spec.js:4]: Good catch! I have missed that.

File: src/v2/view-builder/utils/AuthenticatorUtil.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { loc } from 'okta';
+
+const addSwitchAuthenticatorLink = (appState, links = []) => {
+  if (appState.hasRemediationObject('select-factor-authenticate')) {
+    links.push({
+      'type': 'link',
+      'label':  loc('mfa.switch', 'login'),
+      'name': 'switchFactor',
```
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/v2/view-builder/utils/AuthenticatorUtil.js:8]: this is for V1. the line after is for authenticator: `switchAuthenticator`. maybe we could rename to `signinUsingDifferentMethod` but feels `switchAuthenticator` still fine given it's just naming convention for UI elements but not label.

File: src/v2/models/AppState.js [hunk 11]
```diff
@@ -28,16 +28,19 @@ export default Model.extend({
   },
 
   derived: {
-    factorProfile: {
-      deps: ['factor'],
-      fn (factor = {}) {
-        return factor.profile || {};
+    authenticatorProfile: {
```
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/v2/models/AppState.js:31]: I'd like to tackle it in another PR (given this is already relative big PR) :D
